presto-main/src/main/java/com/facebook/presto/execution/scheduler/SqlQueryScheduler.java

```diff
@@ -641,148 +639,137 @@ public Duration getTotalCpuTime()
-                    }
-                    stageLinkages.get(stage.getStageId())
-                            .processScheduleResults(stage.getState(), result.getNewTasks());
-                    schedulerStats.getSplitsScheduledPerIteration().add(result.getSplitsScheduled());
-                    if (result.getBlockedReason().isPresent()) {
-                        switch (result.getBlockedReason().get()) {
-                            case WRITER_SCALING:
-                                // no-op
-                                break;
-                            case WAITING_FOR_SOURCE:
-                                schedulerStats.getWaitingForSource().update(1);
-                                break;
-                            case SPLIT_QUEUES_FULL:
-                                schedulerStats.getSplitQueuesFull().update(1);
-                                break;
-                            case MIXED_SPLIT_QUEUES_FULL_AND_WAITING_FOR_SOURCE:
-                                schedulerStats.getMixedSplitQueuesFullAndWaitingForSource().update(1);
-                                break;
-                            case NO_ACTIVE_DRIVER_GROUP:
-                                schedulerStats.getNoActiveDriverGroup().update(1);
-                                break;
-                            default:
-                                throw new UnsupportedOperationException("Unknown blocked reason: " + result.getBlockedReason().get());
-                        }
-                    }
+
+            List<ExecutionSchedule> sectionExecutionSchedules = new LinkedList<>();
+
+            while (!Thread.currentThread().isInterrupted()) {
+                // remove finished section
+                sectionExecutionSchedules.removeIf(ExecutionSchedule::isFinished);
+
+                // try to pull more section that are ready to be run
+                List<StreamingPlanSection> sectionsReadyForExecution = getSectionsReadyForExecution();
+
+                // all finished
+                if (sectionsReadyForExecution.isEmpty() && sectionExecutionSchedules.isEmpty()) {
+                    break;
                 }
 
-                // make sure to update stage linkage at least once per loop to catch async state changes (e.g., partial cancel)
-                for (SqlStageExecution stage : stages) {
-                    if (!completedStages.contains(stage.getStageId()) && stage.getState().isDone()) {
+                List<List<SqlStageExecution>> sectionStageExecutions = getStageExections(sectionsReadyForExecution);
+                sectionStageExecutions.forEach(scheduledStages::addAll);
+                sectionStageExecutions.stream()
+                        .map(executionPolicy::createExecutionSchedule)
+                        .forEach(sectionExecutionSchedules::add);
+
+                while (sectionExecutionSchedules.stream().noneMatch(ExecutionSchedule::isFinished)) {
```

Comment:
> So this part (line 687 - line 732) is the same as original code (line 713 - line 752 in old code) ?

Yes, it should've remained the same.

Reply:
> Also curious about the condition for this while loop. So it will break once any section schedule is finished?

Yes, just in case there some other sections become ready for scheduling.